parachain/pallets/outbound-queue/src/lib.rs

```diff
@@ -92,153 +111,184 @@ pub mod pallet {
-				.clone()
-				.into_iter()
-				.map(|msg| ethabi::encode(&vec![msg.into()]))
-				.collect();
+		pub(crate) fn commit_messages() {
+			let messages_count = MessageLeaves::<T>::decode_len().unwrap_or_default() as u64;
+			if messages_count == 0 {
+				return
+			}
 
-			let commitment_hash =
-				merkle_root::<<T as Config>::Hashing, Vec<Vec<u8>>, Vec<u8>>(eth_messages.clone());
+			// Create merkle root of messages
+			let messages_root =
+				merkle_root::<<T as Config>::Hashing, _>(MessageLeaves::<T>::stream_iter());
 
+			// Insert merkle root into the block header
 			<frame_system::Pallet<T>>::deposit_log(sp_runtime::DigestItem::Other(
-				commitment_hash.to_fixed_bytes().into(),
+				messages_root.to_fixed_bytes().into(),
 			));
 
-			Self::deposit_event(Event::Committed {
-				hash: commitment_hash,
-				data: message_queue.to_vec(),
+			Self::deposit_event(Event::MessagesCommitted {
+				root: messages_root,
+				count: messages_count,
 			});
+		}
+
+		fn prepare_enqueued_message(
+			xcm_hash: XcmHash,
+			origin: ParaId,
+			handler: u16,
+			payload: &[u8],
+		) -> Result<BoundedVec<u8, MaxEnqueuedMessageSizeOf<T>>, SubmitError> {
+			// The inner payload should not be too large
+			ensure!(
+				payload.len() < T::MaxMessagePayloadSize::get() as usize,
+				SubmitError::MessageTooLarge
+			);
+			let message: EnqueuedMessage = EnqueuedMessage {
+				xcm_hash,
+				origin: origin.clone(),
+				handler,
+				payload: payload.into(),
+			};
+			// The whole message should not be too large
+			let message = message.encode().try_into().map_err(|_| SubmitError::MessageTooLarge)?;
+			Ok(message)
+		}
+	}
 
-			set(commitment_hash.as_bytes(), &eth_messages.encode());
+	impl<T: Config> OutboundQueueTrait for Pallet<T> {
+		/// Ensure that the message isn't too large
+		fn validate(
+			xcm_hash: XcmHash,
+			origin: ParaId,
+			handler: u16,
+			payload: &[u8],
+		) -> Result<(), SubmitError> {
+			Self::prepare_enqueued_message(xcm_hash, origin, handler, payload).map(|_| ())
+		}
 
-			return T::WeightInfo::on_commit(message_count, average_payload_size)
+		/// Submit message on the outbound channel
+		fn submit(
+			xcm_hash: XcmHash,
+			origin: ParaId,
+			handler: u16,
+			payload: &[u8],
+		) -> Result<(), SubmitError> {
+			let message = Self::prepare_enqueued_message(xcm_hash, origin, handler, payload)?;
+			T::MessageQueue::enqueue_message(
+				message.as_bounded_slice(),
+				AggregateMessageOrigin::Parachain(origin),
+			);
+			Self::deposit_event(Event::MessageQueued { xcm_hash });
+			Ok(())
 		}
+	}
+
+	impl<T: Config> ProcessMessage for Pallet<T> {
+		type Origin = AggregateMessageOrigin;
+		fn process_message(
+			message: &[u8],
+			_: Self::Origin,
+			meter: &mut frame_support::weights::WeightMeter,
+		) -> Result<bool, ProcessMessageError> {
+			ensure!(
+				MessageLeaves::<T>::decode_len().unwrap_or(0) <
+					T::MaxMessagesPerBlock::get() as usize,
+				ProcessMessageError::Yield
+			);
```

Comment:
Is `MaxMessagesPerBlock` hard limit check still nessessary here? `MessageQueue` will just schedule dynamically based on current block limit assume we can provide accurate weight for `do_process_message`

Reply:
I think so or it's not easy to benchmark `on_finalize` based on worst case.

Reply:
yeah, correct 👍